### ISSUE TITLE
'az deployment group create --template-spec' does not work in certain locations

### ISSUE THREAD

**Issue by usrme** [Azure/azure-cli#17733]
## Describe the bug

**Command Name**
`az deployment group create --template-spec`

## To Reproduce

- Create a resource group in a known to be working location, France South for example, with a virtual machine that has Azure CLI installed
- Create a resource group in a broken location, West Europe for example, with a virtual machine that has Azure CLI installed
- Create some kind of template spec in some other resource group
  - Location does not matter for the template spec itself
- From those virtual machines attempt a deployment against those same resource groups, or any other using template specs
- The West Europe one fails, but the France South one succeeds

```console
$ az deployment group create --resource-group rg-westeurope \
  --template-spec /subscriptions/<subscription ID>/resourceGroups/rg-template_specs/providers/Microsoft.Resources/templateSpecs/ts-test/versions/latest \
  --name ts-deployment \
  --debug

<output truncated>
cli.azure.cli.core.util: azure.cli.core.util.handle_exception is called with an exception:
cli.azure.cli.core.util: Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 691, in _run_job
    result = cmd_copy(params)
  File "/usr/local/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 328, in __call__
    return self.handler(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/azure/cli/core/__init__.py", line 807, in default_command_handler
    return op(**command_args)
  File "/usr/local/lib/python3.6/site-packages/azure/cli/command_modules/resource/custom.py", line 544, in deploy_arm_template_at_resource_group
    no_prompt=no_prompt, template_spec=template_spec, query_string=query_string)
  File "/usr/local/lib/python3.6/site-packages/azure/cli/command_modules/resource/custom.py", line 571, in _deploy_arm_template_at_resource_group
    no_prompt=no_prompt, template_spec=template_spec, query_string=query_string)
  File "/usr/local/lib/python3.6/site-packages/azure/cli/command_modules/resource/custom.py", line 896, in _prepare_deployment_properties_unmodified
    template_obj = show_resource(cmd=cmd, resource_ids=[template_spec]).properties['template']
KeyError: 'template'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/azure/cli/core/commands/arm.py", line 102, in handle_template_based_exception
    raise CLIError(ex.inner_exception.error.message)
AttributeError: 'KeyError' object has no attribute 'inner_exception'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/knack/cli.py", line 231, in invoke
    cmd_result = self.invocation.execute(args)
  File "/usr/local/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 657, in execute
    raise ex
  File "/usr/local/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 720, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "/usr/local/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 712, in _run_job
    return cmd_copy.exception_handler(ex)
  File "/usr/local/lib/python3.6/site-packages/azure/cli/core/commands/arm.py", line 104, in handle_template_based_exception
    raise CLIError(ex)
knack.util.CLIError: 'template'

cli.azure.cli.core.azclierror: 'template'
az_command_data_logger: 'template'
cli.knack.cli: Event: Cli.PostExecute [<function AzCliLogging.deinit_cmd_metadata_logging at 0x7f66f225b730>]
az_command_data_logger: exit code: 1
cli.__main__: Command ran in 1.039 seconds (init: 0.188, invoke: 0.851)
telemetry.save: Save telemetry record of length 2983 in cache
telemetry.check: Negative: The /root/.azure/telemetry.txt was modified at 2021-04-16 09:35:28.145237, which in less than 600.000000 s
```

## Expected Behavior

For a valid API version to be returned for the available resource types.

## Environment Summary

```text
Linux-5.11.13-200.fc33.x86_64-x86_64-with-glibc2.32
Python 3.9.2
Installer: RPM

azure-cli 2.22.0
```

## Additional Context

When executing from West Europe there is a `KeyError`, but from France South there isn't and the command finishes successfully. In both cases the only difference is the location from which the request went out. Looking at the debug output in full for both the following can be observed, which requests the registered resources providers and available API versions:

```text
urllib3.connectionpool: Starting new HTTPS connection (1): management.azure.com:443
urllib3.connectionpool: https://management.azure.com:443 "GET /subscriptions/<subscription ID>/providers/Microsoft.Resources?api-version=2020-10-01 HTTP/1.1" 200 2305
```

For **West Europe** the response includes:

```json
{
    "resourceType": "templateSpecs",
    "locations": [
        "East Asia",
        "Southeast Asia",
        "Australia East",
        "Australia Central",
        "Australia Central 2",
        "Australia Southeast",
        "Brazil South",
        "Canada Central",
        "Canada East",
        "Switzerland North",
        "Germany West Central",
        "France South 2",
        "France South",
        "Central US",
        "North Central US",
        "France Central",
        "UK South",
        "UK West",
        "Central India",
        "West India",
        "South India",
        "Japan East",
        "Japan West",
        "Korea Central",
        "Korea South",
        "North Europe",
        "Norway East",
        "UAE North",
        "West Central US",
        "West Europe",
        "West US 2",
        "West US",
        "South Central US",
        "South Africa North"
    ],
    "apiVersions": [
        "2021-03-01-preview",
        "2019-06-01-preview"
    ],
    "capabilities": "CrossResourceGroupResourceMove, CrossSubscriptionResourceMove, SupportsTags, SupportsLocation"
}
```

For France South the `2021-03-01-preview` API version is missing in the response. Thus, when the actual template spec is queried using:

```text
urllib3.connectionpool: https://management.azure.com:443 "GET /subscriptions/<subscription ID>/resourceGroups/rg-template_specs/providers/Microsoft.Resources/templateSpecs/ts-test/versions/latest?api-version=2021-03-01-preview HTTP/1.1" 200 None
```

Then it causes an error because that API version actually isn't valid for West Europe. This can be further confirmed [here](https://docs.microsoft.com/en-us/rest/api/resources/templatespecversions/get#code-try-0) by attempting to create a template spec deployment to West Europe using that API version, which responds with 400:

Headers:

```text
cache-control: no-cache
content-length: 726
content-type: application/json; charset=utf-8
expires: -1
pragma: no-cache
x-ms-correlation-request-id: bf6f21a9-6ce5-4699-bc00-83d48a7eae76
x-ms-failure-cause: gateway
x-ms-request-id: bf6f21a9-6ce5-4699-bc00-83d48a7eae76
x-ms-routing-request-id: FRANCESOUTH:20210416T095723Z:bf6f21a9-6ce5-4699-bc00-83d48a7eae76
```

Body:

```json
{
  "error": {
    "code": "NoRegisteredProviderFound",
    "message": "No registered resource provider found for location 'westeurope' and API version '2021-03-01-preview' for type 'templateSpecs/versions'. The supported api-versions are '2019-06-01-preview'. The supported locations are 'eastasia, southeastasia, australiaeast, australiacentral, australiacentral2, australiasoutheast, brazilsouth, canadacentral, canadaeast, switzerlandnorth, germanywestcentral, eastus2, eastus, centralus, northcentralus, francecentral, uksouth, ukwest, centralindia, westindia, southindia, japaneast, japanwest, koreacentral, koreasouth, northeurope, norwayeast, uaenorth, westcentralus, westeurope, westus2, westus, southcentralus, southafricanorth'."
  }
}
```

**Comment by yonzhan** [comment]
ARM

**Comment by zhoxing-ms** [comment]
@detienne20 could you please help to have a look at this issue~

**Comment by usrme** [comment]
Would it be possible to get an update to this? This is quite a large breaking change in how `az ts` works and is a deciding factor whether alternatives should be considered or not.

**Comment by detienne20** [comment]
@usrme Thank you for patience in this matter, I am thoroughly investigating this issue and will update this ticket as soon as possible.

**Comment by usrme** [comment]
To those also affected by this: I've found a somewhat robust workaround that relies on using `az rest` directly with an additional `wait` added with `az deployment sub wait`.

Here are some steps that worked for me:

- Read in parameters from a parameters file if you rely on parameters: `parameters=$(jq -c '.parameters' parameters.json)`
- Get resource ID of template spec you are referencing
- Call `az rest`:

```
$ az rest --method put \
    --url "https://management.azure.com/subscriptions/$(az account show | \
      jq -r '.id')/resourcegroups/${resource_group_name}/providers/Microsoft.Resources/deployments/${deployment_name}?api-version=2020-10-01" \
    --body "{\"properties\": {\"templateLink\": {\"id\": \"${template_spec_resource_id}\"}, \"parameters\": ${parameters}, \"mode\": \"Incremental\"}}" > /dev/null 2>&1
...
$ az deployment group wait --created --interval 2 --name "$deployment_name" --resource-group "$resource_group_name"
...
```

**Comment by tonystz** [comment]
I found the same issue in UAE North region. and now it block me.

**Comment by lohse614** [comment]
same here in germanywestcentral...
└─ $ ▶ az --version
azure-cli                         2.22.1

core                              2.22.1
telemetry                          1.0.6

Extensions:
portal                             0.1.1
aks-preview                        0.5.5

Python location '/opt/az/bin/python3'
Extensions directory '/home/lohse/.azure/cliextensions'

Python (Linux) 3.6.10 (default, Apr 19 2021, 07:28:54) 
[GCC 9.3.0]

Legal docs and information: aka.ms/AzureCliLegal


Your CLI is up-to-date.

**Comment by fluffy-cakes** [comment]
I have a similar issue with **UK South**; `az deployment sub create -l uksouth -s $myid --debug --verbose`

Last part of `debug` shows the following:
```log
cli.azure.cli.core.util: azure.cli.core.util.handle_exception is called with an exception:
cli.azure.cli.core.util: Traceback (most recent call last):
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 691, in _run_job
    result = cmd_copy(params)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 328, in __call__
    return self.handler(*args, **kwargs)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/__init__.py", line 807, in default_command_handler
    return op(**command_args)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/command_modules/resource/custom.py", line 464, in deploy_arm_template_at_subscription_scope
    no_prompt=no_prompt, template_spec=template_spec, query_string=query_string)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/command_modules/resource/custom.py", line 488, in _deploy_arm_template_at_subscription_scope
    template_spec=template_spec, query_string=query_string)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/command_modules/resource/custom.py", line 896, in _prepare_deployment_properties_unmodified
    template_obj = show_resource(cmd=cmd, resource_ids=[template_spec]).properties['template']
KeyError: 'template'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/arm.py", line 102, in handle_template_based_exception
    raise CLIError(ex.inner_exception.error.message)
AttributeError: 'KeyError' object has no attribute 'inner_exception'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/az/lib/python3.6/site-packages/knack/cli.py", line 231, in invoke
    cmd_result = self.invocation.execute(args)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 657, in execute
    raise ex
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 720, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 712, in _run_job
    return cmd_copy.exception_handler(ex)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/arm.py", line 104, in handle_template_based_exception
    raise CLIError(ex)
knack.util.CLIError: 'template'

cli.azure.cli.core.azclierror: 'template'
az_command_data_logger: 'template'
cli.knack.cli: Event: Cli.PostExecute [<function AzCliLogging.deinit_cmd_metadata_logging at 0x7f24ab7ced08>]
az_command_data_logger: exit code: 1
cli.__main__: Command ran in 1.217 seconds (init: 0.195, invoke: 1.022)
telemetry.save: Save telemetry record of length 3040 in cache
telemetry.check: Negative: The /root/.azure/telemetry.txt was modified at 2021-05-06 09:24:22.096546, which in less than 600.000000 s
```

Details to help
```log
➜  template_spec_test az --version
azure-cli                         2.22.1 *

core                              2.22.1 *
telemetry                          1.0.6

Python location '/opt/az/bin/python3'
Extensions directory '/root/.azure/cliextensions'

Python (Linux) 3.6.10 (default, Apr 19 2021, 07:28:54)
[GCC 9.3.0]

Legal docs and information: aka.ms/AzureCliLegal


You have 2 updates available. Consider updating your CLI installation with 'az upgrade'

Please let us know how we are doing: https://aka.ms/azureclihats
and let us know if you're interested in trying out our newest features: https://aka.ms/CLIUXstudy
➜  template_spec_test python3 --version
Python 3.8.5
```

PowerShell seems to work fine:
```log
paul  template_spec_test  09:35  $id = (Get-AzTemplateSpec -ResourceGroupName "asdf" -Name "qwer" -Version "1.0").Versions.Id

paul  template_spec_test  09:35  $id

/subscriptions/ASDF/resourceGroups/asdf/providers/Microsoft.Resources/templateSpecs/qwer/versions/1.0
paul  template_spec_test  09:35  New-AzSubscriptionDeployment `
>>     -TemplateSpecId $id `
>>     -Location "uksouth" `
>>     -Verbose

VERBOSE:
VERBOSE: 09:35:53 - Template is valid.
VERBOSE: 09:35:54 - Create template deployment 'd2d13661-4f54-4bdf-8c0d-de60c5a2dd55'
VERBOSE: 09:35:54 - Checking deployment status in 5 seconds
VERBOSE: 09:35:59 - Resource Microsoft.Resources/deployments 'asdf' provisioning status is running
VERBOSE: 09:35:59 - Resource Microsoft.Network/virtualNetworks 'asdf' provisioning status is running   
VERBOSE: 09:35:59 - Resource Microsoft.Resources/resourceGroups 'qwer' provisioning status is succeeded
VERBOSE: 09:35:59 - Checking deployment status in 5 seconds
VERBOSE: 09:36:04 - Resource Microsoft.Resources/deployments 'asdf' provisioning status is succeeded
VERBOSE: 09:36:04 - Resource Microsoft.Network/virtualNetworks 'asdf' provisioning status is succeeded

Id                      : /subscriptions/ASDF/providers/Microsoft.Resources/deployments/d2d13661-4f54-4bdf-8c0d-de60c5a2dd55
DeploymentName          : d2d13661-4f54-4bdf-8c0d-de60c5a2dd55
Location                : uksouth
ProvisioningState       : Succeeded
Timestamp               : 06/05/2021 08:36:12     
Mode                    : Incremental
TemplateLink            :
                          Uri            :        
                          ContentVersion : 1.0.0.0

Parameters              : 
Outputs                 : 
DeploymentDebugLogLevel : 
```

**Comment by wiebeck** [comment]
@fluffy-cakes Upgrade to Azure CLI **2.23.0**, the issue has been fixed with #17896. :)

**Comment by fluffy-cakes** [comment]
far out! good reply, i didn't see it got released today, only saw it yesterday still on 2.22.0. cheers 👍🏼

**Comment by fluffy-cakes** [comment]
just re-ran it with new version successfully. thanks!

**Comment by usrme** [comment]
Closing as this particular issue is indeed now fixed with version 2.23.0. Please still have a look at #17759!
